client/src/resources/amount.rs

```diff
@@ -141,23 +142,82 @@ impl<'de> Deserialize<'de> for Amount {
         D: Deserializer<'de>,
     {
         let s = String::deserialize(d)?;
+        let amount: Amount = Amount::from_str(&s).map_err(|e| match e {
+            ParseAmountError::ExceedsDecimalLength(_val) => {
+                de::Error::custom("Amount has too many digits of precision.")
+            }
+            ParseAmountError::IntError(err) => de::Error::custom(err.to_string()),
+        })?;
+        Ok(amount)
+    }
+}
+
+#[cfg(test)]
+mod deserialize_amount_tests {
+    use super::*;
+    use serde_json;
+
+    #[test]
+    fn it_raises_amount_by_ten_million() {
+        let amount: Amount = serde_json::from_str("\"2.12\"").unwrap();
+        assert_eq!(amount, Amount(21_200_000));
+    }
+
+    #[test]
+    fn it_parses_the_smallest_value() {
+        let amount: Amount = serde_json::from_str("\"0.0000001\"").unwrap();
+        assert_eq!(amount, Amount(1));
+    }
+
+    #[test]
+    fn it_handles_integer_strings() {
+        let amount: Amount = serde_json::from_str("\"212\"").unwrap();
+        assert_eq!(amount, Amount(2_120_000_000));
+    }
+
+    #[test]
+    fn it_errors_floats_with_more_than_7_decimals() {
+        let amount = serde_json::from_str::<Amount>("\"0.212847948\"");
+        assert!(amount.is_err());
+    }
+}
+
+/// Recreating ParseIntError since its interior is private
+/// https://doc.rust-lang.org/src/core/num/mod.rs.html#3949
+#[derive(Debug)]
+pub enum ParseAmountError {
+    /// An error describing the case where the amount contains
+    /// too many digits of precision to parse correctly
+    ExceedsDecimalLength(String),
```

Comment:
I don't think we need a string inside this type. I think the type explains what's wrong well enough.